packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1417,6 +1417,84 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
+  /// {@macro flutter.widgets.delegatedTransition}
+  ///
+  /// `receivedTransition` will use the above route's [delegatedTransition] in
+  /// order to show the right route transition when the above route either enters
+  /// or leaves the navigation stack. If not null, the `receivedTransition` will
+  /// wrap the route content through [buildFlexTransitions].
+  DelegatedTransition? receivedTransition;
+
+  /// A method to trigger setting the [delegatedTransition] for a route.
+  ///
+  /// It is possible for a route to supply a different [delegatedTransition]
+  /// depending on the context of the tree at its location. For example Material's
+  /// [ZoomPageTransitionsBuilder] depends on the [Theme] to conditionally provide
+  /// a non-snapshotted [delegatedTransition].
+  ///
+  /// In some cases, the previous route may ask for the [delegatedTransition] of
+  /// a route before it is built. In these cases, the previous route will call
+  /// [setDelegatedTransition] on the the unbuilt route using its own
+  /// [BuildContext]. Override this method to set the [delegatedTransition] based
+  /// on the provided [BuildContext]. This context will be from the PREVIOUS
+  /// route, so do not use logic that requires the [BuildContext] of the current
+  /// route.
+  void setDelegatedTransition(BuildContext context) {}
+
+  /// Wraps the transitions of this route with a [DelegatedTransition], when
+  /// [receivedTransition] is not null.
+  ///
+  /// This allows the current route to sync its exit transition with the
+  /// entrance transition of the incoming route, as well as the exit animation
+  /// when the above route is popped.
+  ///
+  /// {@tool dartpad}
+  /// This sample shows an app that uses three different page transitions, a
+  /// Material Zoom transition, the standard Cupertino sliding transition, and a
+  /// custom vertical transition. All of the page routes are able to inform the
+  /// previous page how to transition off the screen to sync with the new page.
+  ///
+  /// ** See code in examples/api/lib/widgets/routes/flexible_route_transitions.0.dart **
+  /// {@end-tool}
+  ///
+  /// {@tool dartpad}
+  /// This sample shows an app that uses the same transitions as the previous
+  /// sample, this time in a [MaterialApp.router].
+  ///
+  /// ** See code in examples/api/lib/widgets/routes/flexible_route_transitions.1.dart **
+  /// {@end-tool}
+  Widget buildFlexTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    final ProxyAnimation proxyAnimation = ProxyAnimation();
+
+    final Animation<double> flexAnimation = (receivedTransition == null) ?
+      secondaryAnimation : proxyAnimation;
+
+    final Widget originalTransitions = buildTransitions(context, animation, flexAnimation, child);
+
+    if (receivedTransition != null) {
+      return receivedTransition!.builder(context, animation, secondaryAnimation, originalTransitions);
```

Comment:
I vaguely remember I asked this before, but I forgot the conclusion. Why do we apply the receivedTransition on top of original transition? I thought it will complete override the original transition

Reply:
Completely overriding introduced some issues when popping backwards because they forgot their original primary transition. They had to clear their recievedTransition when the route above them finished popping, but quick pops consecutive pops, train hopping, and other complicated situations became tricky. Applying the receivedTransition transition on top of the existing transitions, and suppressing the secondary transition while the receivedTransition was there, solved those issues.

Reply:
IIRC this shows up when navigating during a transition (say popping and then popping again before the first pop transition ends). This way we can keep the widget tree the same and smoothly go from one transition to another.

I think it's probably similar to why buildTransitions produces a widget tree with both the `animation` and `secondaryAnimation` widgets in it at the same time.

Reply:
sounds good